// gantt/es/gantt/DayHeader.js
import h from '../h';
import { getDates } from '../utils';
import YearMonth from './YearMonth';
export default function DayHeader(_ref) {
  var styles = _ref.styles,
      unit = _ref.unit,
      minTime = _ref.minTime,
      maxTime = _ref.maxTime,
      height = _ref.height,
      offsetY = _ref.offsetY,
      maxTextWidth = _ref.maxTextWidth;
  var dates = getDates(minTime, maxTime);
  var ticks = [];
  var x0 = maxTextWidth;
  var y0 = offsetY / 2;
  var RH = height - y0;
  var len = dates.length - 1;

  for (var i = 0; i < len; i++) {
    var cur = new Date(dates[i]);
    var day = cur.getDay();
    var x = x0 + (dates[i] - minTime) / unit;
    var t = (dates[i + 1] - dates[i]) / unit;
    ticks.push(h("g", null, day === 0 || day === 6 ? h("rect", {
      x: x,
      y: y0,
      width: t,
      height: RH,
      style: styles.week
    }) : null, h("line", {
      x1: x,
      x2: x,
      y1: y0,
      y2: offsetY,
      style: styles.line
    }), h("text", {
      x: x + t / 2,
      y: offsetY * 0.75,
      style: styles.text3
    }, cur.getDate()), i === len - 1 ? h("line", {
      x1: x + t,
      x2: x + t,
      y1: y0,
      y2: offsetY,
      style: styles.line
    }) : null));
  }

  return h("g", null, h(YearMonth, {
    styles: styles,
    unit: unit,
    dates: dates,
    offsetY: offsetY,
    minTime: minTime,
    maxTime: maxTime,
    maxTextWidth: maxTextWidth
  }), ticks);
}
