// gantt/es/gantt/WeekHeader.js
import h from '../h';
import { getDates, addDays, DAY } from '../utils';
import YearMonth from './YearMonth';
export default function WeekHeader(_ref) {
  var styles = _ref.styles,
      unit = _ref.unit,
      minTime = _ref.minTime,
      maxTime = _ref.maxTime,
      height = _ref.height,
      offsetY = _ref.offsetY,
      maxTextWidth = _ref.maxTextWidth;
  var dates = getDates(minTime, maxTime);
  var weeks = dates.filter(function (v) {
    return new Date(v).getDay() === 0;
  });
  weeks.push(maxTime);
  var ticks = [];
  var x0 = maxTextWidth;
  var y0 = offsetY;
  var RH = height - y0;
  var d = DAY / unit;
  var len = weeks.length - 1;

  for (var i = 0; i < len; i++) {
    var cur = new Date(weeks[i]);
    var x = x0 + (weeks[i] - minTime) / unit;
    var curDay = cur.getDate();
    var prevDay = addDays(cur, -1).getDate();
    ticks.push(h("g", null, h("rect", {
      x: x - d,
      y: y0,
      width: d * 2,
      height: RH,
      style: styles.week
    }), h("line", {
      x1: x,
      x2: x,
      y1: offsetY / 2,
      y2: offsetY,
      style: styles.line
    }), h("text", {
      x: x + 3,
      y: offsetY * 0.75,
      style: styles.text2
    }, curDay), x - x0 > 28 ? h("text", {
      x: x - 3,
      y: offsetY * 0.75,
      style: styles.text1
    }, prevDay) : null));
  }

  return h("g", null, h(YearMonth, {
    styles: styles,
    unit: unit,
    dates: dates,
    offsetY: offsetY,
    minTime: minTime,
    maxTime: maxTime,
    maxTextWidth: maxTextWidth
  }), ticks);
}
