// gantt/lib/gantt/DayHeader.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = DayHeader;

var _h = _interopRequireDefault(require("../h"));

var _utils = require("../utils");

var _YearMonth = _interopRequireDefault(require("./YearMonth"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function DayHeader(_ref) {
  var styles = _ref.styles,
      unit = _ref.unit,
      minTime = _ref.minTime,
      maxTime = _ref.maxTime,
      height = _ref.height,
      offsetY = _ref.offsetY,
      maxTextWidth = _ref.maxTextWidth;
  var dates = (0, _utils.getDates)(minTime, maxTime);
  var ticks = [];
  var x0 = maxTextWidth;
  var y0 = offsetY / 2;
  var RH = height - y0;
  var len = dates.length - 1;

  for (var i = 0; i < len; i++) {
    var cur = new Date(dates[i]);
    var day = cur.getDay();
    var x = x0 + (dates[i] - minTime) / unit;
    var t = (dates[i + 1] - dates[i]) / unit;
    ticks.push((0, _h["default"])("g", null, day === 0 || day === 6 ? (0, _h["default"])("rect", {
      x: x,
      y: y0,
      width: t,
      height: RH,
      style: styles.week
    }) : null, (0, _h["default"])("line", {
      x1: x,
      x2: x,
      y1: y0,
      y2: offsetY,
      style: styles.line
    }), (0, _h["default"])("text", {
      x: x + t / 2,
      y: offsetY * 0.75,
      style: styles.text3
    }, cur.getDate()), i === len - 1 ? (0, _h["default"])("line", {
      x1: x + t,
      x2: x + t,
      y1: y0,
      y2: offsetY,
      style: styles.line
    }) : null));
  }

  return (0, _h["default"])("g", null, (0, _h["default"])(_YearMonth["default"], {
    styles: styles,
    unit: unit,
    dates: dates,
    offsetY: offsetY,
    minTime: minTime,
    maxTime: maxTime,
    maxTextWidth: maxTextWidth
  }), ticks);
}
