// gantt/lib/gantt/MonthHeader.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = MonthHeader;

var _h = _interopRequireDefault(require("../h"));

var _utils = require("../utils");

var _Year = _interopRequireDefault(require("./Year"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function MonthHeader(_ref) {
  var styles = _ref.styles,
      unit = _ref.unit,
      minTime = _ref.minTime,
      maxTime = _ref.maxTime,
      offsetY = _ref.offsetY,
      maxTextWidth = _ref.maxTextWidth;
  var MONTH = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];
  var dates = (0, _utils.getDates)(minTime, maxTime);
  var months = dates.filter(function (v) {
    return new Date(v).getDate() === 1;
  });
  months.unshift(minTime);
  months.push(maxTime);
  var ticks = [];
  var x0 = maxTextWidth;
  var y0 = offsetY / 2;
  var len = months.length - 1;

  for (var i = 0; i < len; i++) {
    var cur = new Date(months[i]);
    var month = cur.getMonth();
    var x = x0 + (months[i] - minTime) / unit;
    var t = (months[i + 1] - months[i]) / unit;
    ticks.push((0, _h["default"])("g", null, (0, _h["default"])("line", {
      x1: x,
      x2: x,
      y1: y0,
      y2: offsetY,
      style: styles.line
    }), t > 30 ? (0, _h["default"])("text", {
      x: x + t / 2,
      y: offsetY * 0.75,
      style: styles.text3
    }, MONTH[month]) : null));
  }

  return (0, _h["default"])("g", null, (0, _h["default"])(_Year["default"], {
    styles: styles,
    unit: unit,
    months: months,
    offsetY: offsetY,
    minTime: minTime,
    maxTime: maxTime,
    maxTextWidth: maxTextWidth
  }), ticks);
}
