// gantt/lib/gantt/YearMonth.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = YearMonth;

var _h = _interopRequireDefault(require("../h"));

var _utils = require("../utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function YearMonth(_ref) {
  var styles = _ref.styles,
      dates = _ref.dates,
      unit = _ref.unit,
      offsetY = _ref.offsetY,
      minTime = _ref.minTime,
      maxTime = _ref.maxTime,
      maxTextWidth = _ref.maxTextWidth;
  var months = dates.filter(function (v) {
    return new Date(v).getDate() === 1;
  });
  months.unshift(minTime);
  months.push(maxTime);
  var ticks = [];
  var x0 = maxTextWidth;
  var y2 = offsetY / 2;
  var len = months.length - 1;

  for (var i = 0; i < len; i++) {
    var cur = new Date(months[i]);
    var str = (0, _utils.formatMonth)(cur);
    var x = x0 + (months[i] - minTime) / unit;
    var t = (months[i + 1] - months[i]) / unit;
    ticks.push((0, _h["default"])("g", null, (0, _h["default"])("line", {
      x1: x,
      x2: x,
      y1: 0,
      y2: y2,
      style: styles.line
    }), t > 50 ? (0, _h["default"])("text", {
      x: x + t / 2,
      y: offsetY * 0.25,
      style: styles.text3
    }, str) : null));
  }

  return (0, _h["default"])("g", null, ticks);
}
