// gantt/lib/gantt/Year.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = Year;

var _h = _interopRequireDefault(require("../h"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function Year(_ref) {
  var styles = _ref.styles,
      months = _ref.months,
      unit = _ref.unit,
      offsetY = _ref.offsetY,
      minTime = _ref.minTime,
      maxTime = _ref.maxTime,
      maxTextWidth = _ref.maxTextWidth;
  var years = months.filter(function (v) {
    return new Date(v).getMonth() === 0;
  });
  years.unshift(minTime);
  years.push(maxTime);
  var ticks = [];
  var x0 = maxTextWidth;
  var y2 = offsetY / 2;
  var len = years.length - 1;

  for (var i = 0; i < len; i++) {
    var cur = new Date(years[i]);
    var x = x0 + (years[i] - minTime) / unit;
    var t = (years[i + 1] - years[i]) / unit;
    ticks.push((0, _h["default"])("g", null, (0, _h["default"])("line", {
      x1: x,
      x2: x,
      y1: 0,
      y2: y2,
      style: styles.line
    }), t > 35 ? (0, _h["default"])("text", {
      x: x + t / 2,
      y: offsetY * 0.25,
      style: styles.text3
    }, cur.getFullYear()) : null));
  }

  return (0, _h["default"])("g", null, ticks);
}
